src/NServiceBus.Core/Routing/EndpointInstances.cs

```diff
@@ -10,47 +10,55 @@ namespace NServiceBus.Routing
     /// </summary>
     public class EndpointInstances
     {
-        internal async Task<IEnumerable<EndpointInstance>> FindInstances(EndpointName endpoint)
+        internal async Task<IEnumerable<EndpointInstance>> FindInstances(string endpoint)
         {
-            var instances = new List<EndpointInstance>();
+            var instances = new HashSet<EndpointInstance>();
             foreach (var rule in rules)
             {
-                instances.AddRange(await rule.Invoke(endpoint).ConfigureAwait(false));
+                var instancesFromRule = await rule.Invoke(endpoint).ConfigureAwait(false);
+                foreach (var instance in instancesFromRule)
+                {
+                    instances.Add(instance);
+                }
             }
-            var distinctInstances = instances.Distinct().ToArray();
-            return distinctInstances.EnsureNonEmpty(() => new EndpointInstance(endpoint));
+            return instances.EnsureNonEmpty(() => new EndpointInstance(endpoint));
         }
 
 
         /// <summary>
         /// Adds a dynamic rule for determining endpoint instances.
         /// </summary>
         /// <param name="dynamicRule">The rule.</param>
-        public void AddDynamic(Func<EndpointName, Task<IEnumerable<EndpointInstance>>> dynamicRule)
+        public void AddDynamic(Func<string, Task<IEnumerable<EndpointInstance>>> dynamicRule)
         {
             rules.Add(dynamicRule);
         }
 
         /// <summary>
         /// Adds static information about an endpoint.
         /// </summary>
-        /// <param name="endpoint">Name of the endpoint.</param>
         /// <param name="instances">A static list of endpoint's instances.</param>
-        public void Add(EndpointName endpoint, params EndpointInstance[] instances)
+        public void Add(params EndpointInstance[] instances) => Add((IEnumerable<EndpointInstance>)instances);
+
+        /// <summary>
+        /// Adds static information about an endpoint.
+        /// </summary>
+        /// <param name="instances">A static list of endpoint's instances.</param>
+        public void Add(IEnumerable<EndpointInstance> instances)
         {
-            Guard.AgainstNull(nameof(endpoint), endpoint);
-            if (instances.Length == 0)
+            if (!instances.Any())
             {
                 throw new ArgumentException("The list of instances can't be empty.", nameof(instances));
             }
+            var endpoint = instances.First().Endpoint;
             if (instances.Any(i => i.Endpoint != endpoint))
             {
-                throw new ArgumentException("At least one of the instances belongs to a different endpoint than specified in the 'endpoint' parameter.", nameof(instances));
+                throw new ArgumentException("The instances belong to different endpoints. The endpoint names do not match.", nameof(instances));
```

Comment:
I think this is somehow weird. Since EndpointInstance contains the logical endpoint information, is there any reason you're unable to use `config.Add(new EndpointInstance(B, "1"), new EndpointInstance(A, "1"))` instead of making two calls to Add then?